### PR TITLE
fix: create lock dir parent if missing

### LINKED ISSUE #10866: pkg: `dune pkg lock "dev/dune.lock"` fails, because `dune.lock` is a sub-directory.
## Expected Behavior
Do not crash

## Actual Behavior
```shell
+|  Internal error, please report upstream including the contents of _build/log.
+|  Description:
+|    ("[Temp.create_temp_dir] called in a nonexistent directory",
+|    { dir = In_source_tree "dev/dune_3b32f9_lock" })
+|  Raised at Stdune__Result.ok_exn in file "otherlibs/stdune/src/result.ml",
+|    line 26, characters 15-22
+|  Called from Dune_pkg__Lock_dir.Write_disk.prepare.build in file
+|    "src/dune_pkg/lock_dir.ml", line 584, characters 26-53
+|  Called from Stdlib__List.iter in file "list.ml", line 110, characters 12-15
+|  Called from Fiber__Core.O.(>>|).(fun) in file "vendor/fiber/src/core.ml",
+|    line 253, characters 36-41
+|  Called from Fiber__Scheduler.exec in file "vendor/fiber/src/scheduler.ml",
+|    line 76, characters 8-11
+|  
+|  I must not crash.  Uncertainty is the mind-killer. Exceptions are the
+|  little-death that brings total obliteration.  I will fully express my cases. 
+|  Execution will pass over me and through me.  And when it has gone past, I
+|  will unwind the stack along its path.  Where the cases are handled there will
+|  be nothing.  Only I will remain.
+|  [1]
```

## Reproduction
What the reproduction does is:
1. Apply the patch
2. And run the test
 
```shell
git apply <<'EOF'
diff --git a/test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety.t/run.t b/test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety.t/run.t
index d952687e8..093ef0f39 100644
--- a/test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety.t/run.t
+++ b/test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety.t/run.t
@@ -4,10 +4,14 @@ Create a lock directory that didn't originally exist
   > (lang dune 3.10)
   > (lock_dir
   >  (repositories mock))
+  > (lock_dir
+  >  (path "dev/dune.lock")
+  >  (repositories mock))
   > (repository
   >  (name mock)
   >  (source "file://$(pwd)/mock-opam-repository"))
   > EOF
+  $ dune pkg lock "dev/dune.lock"
   $ dune pkg lock
   Solution for dune.lock:
   (no dependencies to lock)
EOF
dune build @test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety
```
## Additional information 
This is coming from the https://github.com/ocaml/dune/pull/10852, the parent directory was not created before creating the temporary directory.

### PR DESCRIPTION
**Pull request by maiste**
This PR fixes #10866. It creates the lock dir if it is missing.

### CODE REVIEW

File: test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety.t/run.t [hunk 1]
```diff
@@ -4,10 +4,16 @@ Create a lock directory that didn't originally exist
...
+  > (lock_dir
+  >  (path "dev/dune.lock")
+  >  (repositories mock))
   > (repository
   >  (name mock)
   >  (source "file://$(pwd)/mock-opam-repository"))
   > EOF
+  $ dune pkg lock "dev/dune.lock"
+  Solution for dev/dune.lock:
+  (no dependencies to lock)
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety.t/run.t:16]: Did this test not fail? Or did you just forget to promote it?

Comment (maiste) [reply test/blackbox-tests/test-cases/pkg/lock-directory-regeneration-safety.t/run.t:16]: My bad: I've mixed the failing test with the promote one... I'll split them in two :+1:

**Comment by maiste** [comment]
I have split the test in two commits: one before the fix and the next one after the fix is applied.

**Review by moyodiallo — APPROVED** [review]
LGTM, thanks.

**Pull request merged**